Compilation of wrong types should not proceed

Try this:

```
let myAdd: (baseValue:number, increment:number) => number =
    function(x, y) { return x + y; };

exports.myAdd = myAdd;
exports.zzz = myAdd(5, '2');
```

It seems to go through. Not sure why.

When I built it with TypeScript, I have this error :

```bash
$ tsc tmp.ts
tmp.ts(4,1): error TS2304: Cannot find name 'exports'.
tmp.ts(5,1): error TS2304: Cannot find name 'exports'.
tmp.ts(5,24): error TS2345: Argument of type 'string' is not assignable to parameter of type 'number'.
```

--- a/index.js
+++ b/index.js
@@ -1,5 +1,6 @@
 /*global __dirname */
 'use strict';
+const transpileModule = require('./transpile');
 const ts = require('typescript');
 const anymatch = require('anymatch');
 const path = require('path');
@@ -34,6 +35,22 @@ const getTsconfig = (root) => {
   return tsconf.compilerOptions || {};
 };
 
+
+const findLessOrEqual = (haystack, needle) => {
+  let i = 0;
+  while (i + 1 < haystack.length && needle >= haystack[i + 1]) {
+    i += 1;
+  }
+  return i === haystack.length ? -1 : i;
+}
+
+const errPos = err => {
+  const lineIndex = findLessOrEqual(err.file.lineMap, err.start);
+  return `Line: ${lineIndex + 1}, Col: ${err.start - err.file.lineMap[lineIndex] + 1}`;
+};
+
+const toMeaningfulMessage = err => `Error ${err.code}: ${err.messageText} (${errPos(err)})`;
+
 class TypeScriptCompiler {
   constructor(config) {
     if (!config) config = {};
@@ -49,6 +66,7 @@ class TypeScriptCompiler {
     this.options.jsx = resolveEnum(this.options.jsx, ts.JsxEmit);
     this.options.emitDecoratorMetadata = this.options.emitDecoratorMetadata !== false,
     this.options.experimentalDecorators = this.options.experimentalDecorators !== false,
+    this.options.noEmitOnError = false; // This can't be true when compiling this way.
     this.options.sourceMap = !!config.sourceMaps;
     this.isIgnored = anymatch(options.ignore || /^(bower_components|vendor|node_modules)/);
     if (this.options.pattern) {
@@ -63,13 +81,17 @@ class TypeScriptCompiler {
     }
     let tsOptions = {
       fileName: params.path,
+      reportDiagnostics: true,
       compilerOptions: this.options
     }
 
     return new Promise((resolve, reject) => {
       let compiled;
       try {
-        compiled = ts.transpileModule(params.data, tsOptions);
+        compiled = transpileModule(params.data, tsOptions);
+        if (compiled.diagnostics.length) {
+          reject(compiled.diagnostics.map(toMeaningfulMessage).join('\n'));
+        }
       } catch (err) {
         return reject(err);
       }

--- a/transpile.js
+++ b/transpile.js
@@ -1,0 +1,64 @@
+'use strict';
+const ts = require('typescript');
+
+const filterErrors = err => err.code !== 1208;
+
+module.exports = function transpileModule(input, transpileOptions) {
+    var options = transpileOptions.compilerOptions ? ts.clone(transpileOptions.compilerOptions) : getDefaultCompilerOptions();
+    options.isolatedModules = true;
+    // transpileModule does not write anything to disk so there is no need to verify that there are no conflicts between input and output paths. 
+    options.suppressOutputPathCheck = true;
+    // Filename can be non-ts file.
+    options.allowNonTsExtensions = true;
+    // We are not returning a sourceFile for lib file when asked by the program,
+    // so pass --noLib to avoid reporting a file not found error.
+    options.noLib = true;
+    // We are not doing a full typecheck, we are not resolving the whole context,
+    // so pass --noResolve to avoid reporting missing file errors.
+    options.noResolve = true;
+    // if jsx is specified then treat file as .tsx
+    var inputFileName = transpileOptions.fileName || (options.jsx ? "module.tsx" : "module.ts");
+    var sourceFile = ts.createSourceFile(inputFileName, input, options.target);
+    if (transpileOptions.moduleName) {
+        sourceFile.moduleName = transpileOptions.moduleName;
+    }
+    sourceFile.renamedDependencies = transpileOptions.renamedDependencies;
+    var newLine = ts.getNewLineCharacter(options);
+    // Output
+    var outputText;
+    var sourceMapText;
+    // Create a compilerHost object to allow the compiler to read and write files
+    var compilerHost = {
+        getSourceFile: function (fileName, target) { return fileName === ts.normalizeSlashes(inputFileName) ? sourceFile : undefined; },
+        writeFile: function (name, text, writeByteOrderMark) {
+            if (ts.fileExtensionIs(name, ".map")) {
+                ts.Debug.assert(sourceMapText === undefined, "Unexpected multiple source map outputs for the file '" + name + "'");
+                sourceMapText = text;
+            }
+            else {
+                ts.Debug.assert(outputText === undefined, "Unexpected multiple outputs for the file: '" + name + "'");
+                outputText = text;
+            }
+        },
+        getDefaultLibFileName: function () { return "lib.d.ts"; },
+        useCaseSensitiveFileNames: function () { return false; },
+        getCanonicalFileName: function (fileName) { return fileName; },
+        getCurrentDirectory: function () { return ""; },
+        getNewLine: function () { return newLine; },
+        fileExists: function (fileName) { return fileName === inputFileName; },
+        readFile: function (fileName) { return ""; },
+        directoryExists: function (directoryExists) { return true; }
+    };
+    var program = ts.createProgram([inputFileName], options, compilerHost);
+    var diagnostics;
+    if (transpileOptions.reportDiagnostics) {
+        diagnostics = [];
+        ts.addRange(/*to*/ diagnostics, /*from*/ program.getSyntacticDiagnostics(sourceFile));
+        ts.addRange(/*to*/ diagnostics, /*from*/ program.getSemanticDiagnostics(sourceFile).filter(filterErrors));
+        ts.addRange(/*to*/ diagnostics, /*from*/ program.getOptionsDiagnostics());
+    }
+    // Emit
+    program.emit();
+    ts.Debug.assert(outputText !== undefined, "Output generation failed");
+    return { outputText: outputText, diagnostics: diagnostics, sourceMapText: sourceMapText };
+}

